src/init.cpp

```diff
@@ -205,7 +205,14 @@ void InitContext(NodeContext& node)
     g_shutdown.emplace();
 
     node.args = &gArgs;
-    node.shutdown = &*g_shutdown;
+    node.shutdown_signal = &*g_shutdown;
+    node.shutdown_request = [&node] {
+        assert(node.shutdown_signal);
+        if (!(*node.shutdown_signal)()) return false;
+        // Wake any threads that may be waiting for the tip to change.
```

Comment:
To avoid a race, this needs to lock `kernel_notifications.m_tip_block_mutex` before calling notify. `(*node.shutdown_signal)()` doesn't necessarily need to happen under the lock, but a lock is required after the signal in order to synchronize. Otherwise the update could be missed and we could potentially wait forever.

This is currently an example of [this trap](https://www.modernescpp.com/index.php/c-core-guidelines-be-aware-of-the-traps-of-condition-variables/) (see the "An atomic predicate" example).

See another example (and solution) [here](https://developers.redhat.com/articles/2022/12/06/implementing-c20-atomic-waiting-libstdc#handling_the_various_type_and_platform_choices) (ctrl+f for "acquisition above is significant").

So this would become:
```c++
if (!(*node.shutdown_signal)()) return false;
// Wake any threads that may be waiting for the tip to change.
if (node.notifications) {
    {
        // synchronize the shutdown state
        LOCK(node.notifications->m_tip_block_mutex);
    }
    node.notifications->m_tip_block_cv.notify_all();
}
```

Reply:
Thanks, good catch. I knew that pthreads required the lock to be held while notifying to avoid the lost wakeup bug (["if predictable scheduling behavior is required"](https://linux.die.net/man/3/pthread_cond_broadcast)), and that c++ dropped this requirement, even recommending against it saying ["Doing so may be a pessimization..."](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_all).

But I didn't know that c++ still requires the lock to be held *before* notifying, even though it now recommends against holding the lock *while* notifying.

It does make sense that after the notifying thread updates the shutdown state, some extra synchronization is needed to make sure the notification is not sent too early, before the waiting thread is actually waiting.

Suggested fix looks right to me. And it looks like this bug might have been present even before this commit, though was much less likely to happen since notify_all call was happening long after the shutdown flag was set, not immediately after.

Reply:
Yeah, I think the "Doing so may be a pessimization..." language encourages this footgun, sadly.

See [here](https://github.com/isocpp/CppCoreGuidelines/issues/1272) for a discussion about inverting the language.